design/src/rfcs/rfc0020_service_builder.md

```diff
@@ -283,4 +280,269 @@ let new_routes = old_routes
+
+There are two points at which the customer might want to apply middleware: around `tower::Service<{Operation}Input, Response = {Operation}Output>` and `tower::Service<http::Response, Response = http::Response>`, that is, before and after the serialization/deserialization is performed.
+
+The three use cases described above are supported by `axum` by virtue of the [Router::route](https://docs.rs/axum/latest/axum/routing/struct.Router.html#method.route) method accepting a `tower::Service`. The reader should consider a similar approach where the service builder setters accept a `tower::Service<http::Request, Response = http::Response>` rather than the `Handler`. In this way the we separate the building of the entire service from the building of individual operations, this leads us to one of the central tenets of this proposal: operations as middleware. More concretely, the customer is provided with the following generated code
+
+```rust
+struct Operation0<S> {
+    inner: S,
+}
+
+impl<S> Service<http::Request> for Operation0<S>
+where
+    S: Service<Operation0Input, Response = Operation0Output, Error = Infallible>
+{
+    type Response = http::Response;
+    type Error = Infallible;
+
+    fn poll_ready(&mut self, cx: &mut Context) -> Poll<Result<(), Self::Error>> {
+        // We defer to the inner service for readiness
+        self.inner.poll_ready(cx)
+    }
+
+    async fn call(&mut self, request: http::Request) -> Result<Self::Response, Self::Error> {
+        let input = /* Create `Operation0Input` from `request: http::Request` */;
+
+        self.inner.call(input).await;
+
+        let response = /* Create `http::Response` from `output: Operation0Output` */
+        response
+    }
+}
+
+fn make_operation_0<S>(inner: S) -> Operation0<S> {
+    Operation0 {
+        inner
+    }
+}
+```
+
+which is concerned with all the same model aware serialization/deserialization we noted in [Handler](#handler). The API usage then becomes
+
+```rust
+async fn handler0(input: Operation0Input) -> Operation0Output {
+    todo!()
+}
+
+async fn handler1(input: Operation1Input, state: State) -> Operation1Output {
+    todo!()
+}
+
+// Model agnostic `into_service` method which transforms a handler `{Operation}Input -> {Operation}Output` to
+// `tower::Service<{Operation}Input, Response = {Operation}Output, Error = Infallible>`
+let svc = OperationHandler::new(handler0);
+
+// Middleware can be applied at this point
+let svc = model_layer.layer(svc);
+
+// Model aware conversion between `tower::Service<{Operation}Input, Response = {Operation}Output, Error = Infallible>`
+// and `tower::Service<http::Request, Response = http::Response, Error = Infallible>`
+let operation0 = make_operation0(svc);
+
+// Middleware can be applied at this point
+let operation0 = http_layer.layer(op1_svc);
+
+OperationRegistryBuilder::default()
+    .operation0(operation0)
```

Comment:
What is preventing the user from passing in an arbitrary `tower::Service` for `operation0` here? The current operation registry ensures that the provided handler takes in and returns the appropriate input and output types.

Reply:
Nothing, it's possible that the customer can pass a `tower::Service<http::Request, Response = http::Response>` here which completely ignores the model. I'm unsure if this is any worse than allowing the customer to apply arbitrary `Layer`s.

Approach C has been added which is more attractive from this POV.